Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -293,162 +344,314 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
+  const auto savefiles = Memcard::GetSavefiles(*memcard, selected_indices);
+  if (savefiles.empty())
+  {
+    ModalMessageBox::warning(this, tr("Export Failed"),
+                             tr("Failed to read selected savefile(s) from memory card."));
+    return;
+  }
 
-  if (result != Memcard::GCMemcardImportFileRetVal::SUCCESS)
+  std::string extension = Memcard::GetDefaultExtension(format);
+
+  if (savefiles.size() == 1)
   {
-    ModalMessageBox::critical(this, tr("Import failed"), tr("Failed to import \"%1\".").arg(path));
+    // when exporting a single save file, let user specify exact path
+    const std::string basename = Memcard::GenerateFilename(savefiles[0].dir_entry);
+    const QString qformatdesc = GetFormatDescription(format);
+    const std::string default_path =
+        fmt::format("{}/{}{}", File::GetUserPath(D_GCUSER_IDX), basename, extension);
+    const QString qfilename = QFileDialog::getSaveFileName(
+        this, tr("Export Save File"), QString::fromStdString(default_path),
+        QStringLiteral("%1 (*%2);;%3 (*)")
+            .arg(qformatdesc, QString::fromStdString(extension), tr("All Files")));
+    if (qfilename.isEmpty())
+      return;
+
+    const std::string filename = qfilename.toStdString();
+    if (!Memcard::WriteSavefile(filename, savefiles[0], format))
+    {
+      File::Delete(filename);
+      ModalMessageBox::warning(this, tr("Export Failed"), tr("Failed to write savefile to disk."));
+    }
+
     return;
   }
 
-  if (!m_slot_memcard[m_active_slot]->Save())
-    PanicAlertFmtT("File write failed");
+  const QString qdirpath =
+      QFileDialog::getExistingDirectory(this, QObject::tr("Export Save Files"),
+                                        QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)));
+  if (qdirpath.isEmpty())
+    return;
 
-  UpdateSlotTable(m_active_slot);
+  const std::string dirpath = qdirpath.toStdString();
+  size_t failures = 0;
+  for (const auto& savefile : savefiles)
+  {
+    // find a free filename so we don't overwrite anything
+    const std::string basepath = dirpath + DIR_SEP + Memcard::GenerateFilename(savefile.dir_entry);
+    std::string filename = basepath + extension;
+    if (File::Exists(filename))
+    {
+      size_t tmp = 0;
+      std::string free_name;
+      do
+      {
+        free_name = fmt::format("{}_{}{}", basepath, tmp, extension);
+        ++tmp;
+      } while (File::Exists(free_name));
+      filename = free_name;
+    }
+
+    if (!Memcard::WriteSavefile(filename, savefile, format))
+    {
+      File::Delete(filename);
+      ++failures;
+    }
+  }
+
+  if (failures > 0)
+  {
+    QString failure_string =
+        tr("Failed to export %n save file(s).", "", static_cast<int>(failures));
+    if (failures == savefiles.size())
+    {
+      ModalMessageBox::warning(this, tr("Export Failed"), failure_string);
+    }
+    else
+    {
+      QString success_string = tr("Successfully exported %n save file(s).", "",
```

Comment:
Would it be worthwhile to put the total here too?

Reply:
Sure, why not.